Config as code/ongoing updates (#1597)
OctopusDeploy/docs: docs/projects/version-control/creating-release-from-a-build-server-plug-in.md

```
$[coalesce(variables['system.pullRequest.sourceCommitId'], variables['build.sourceVersion'])]
```

**Note:** this approach doesn't populate the commit details for manually triggered runs. It is recommended that you provide the values for both branch and commit in this case.

## Github Actions

Github Actions provides different event data based on if the build is triggered by a branch push versus via a Pull Request (PR). You need to use a conditional statement to select the right variable. Here's a suggested method to do so.

For *Git Reference*, use
```
${{ github.head_ref || github.ref }}
```

For *Git Commit*, use
```
${{ github.event.push.after || github.event.pull_request.head.sha }}
```

**Note:** this approach doesn't populate the commit details for manually triggered runs. It is recommended that you provide the values for both branch and commit in this case.

## TeamCity

TeamCity provides different data based on if the build is triggered by a branch push versus via a Pull Request (PR). There is no easy way to automate selecting the right one. Based on your process, we suggest using one of the following options.

### Build triggered by a branch push

>  For *Git Reference*, use 

Reviewer: Just to further clarify for TC there are 2 complexities,
**1. Is this build on triggered directly on the branch or via a Pull request.** In this case they need to use the different pull-request specific parameter to get the gitref (commit is unchanged). If we end up supporting pull request git refs natively like we do tags this might go away completely.
**2. Is this build the original source of the VCS root or is it a chained build?** - Often it is the simple case with one repo the passed along anyway and so the below `dep....` format isn't required. We could simplify the docs and leave this out since referencing variables from the previous steps is arguable something that they should know about TC.

Just to give you more context on what that with some examples from our builds:

### Commit
For a build triggered where the vcs root is the repository that contains your OCL, you can use the `build.vcs.number` (or the fully qualified parameter that includes the VCS ID `build.vcs.number.<Vcs_Root_Id>` if you have multiple VCS sources used in a build.)
![image](https://user-images.githubusercontent.com/1830666/180696661-a47b4497-acf7-4026-8dca-954be65a532f.png)

----
If however the build is the result of a chained step you may need to reference a parameter from the original build dependency (in this case the dependancy is the project example above). 
i.e. `dep.<Dependant_Build_ID>.build.vcs.number` or `dep.<Dependant_Build_ID>.build.vcs.number.<Vcs_Root_Id>`
![image](https://user-images.githubusercontent.com/1830666/180696496-df119545-0fca-4cff-87d4-f434461add31.png)

### GitRef
Similarly, for the `GitRef` you can reference `teamcity.build.branch`  (or `teamcity.build.vcs.branch.<Vcs_Root_Id>`).
![image](https://user-images.githubusercontent.com/1830666/180697816-b4bd2481-8ffd-4340-ac20-a404b3926a65.png) (or similarly the `dep.<Dependant_Build_ID.teamcity.build.branch` as described above for a chained build)

If however used in a PullRequest, the variable will need to be `teamcity.pullRequest.source.branch` 
![image](https://user-images.githubusercontent.com/1830666/180697767-fc56e43d-1511-4772-935d-60fae5a55d28.png)

Author: Thank you! Incorporated the changes in :-)

---
title: Creating releases from a build server plugin on a version-controlled project
description: Examples of how to ensure that the right branch is used to create the release when using a build server plugin.
position: 45 
---

!include <build-server-plugin-version-control-fields>

Below are some examples of how to auto populate these fields using build parameters for common build servers.

## Azure DevOps
 
Azure DevOps stores information differently based on if the build is triggered by a branch push versus via a Pull Request (PR). You need to use a conditional statement to select the right variable. Here's a suggested method to do so.

For *Git Reference*, use
```
$[coalesce(variables['system.pullRequest.sourceBranch'], variables['build.sourceBranch'])]
```

For *Git Commit*, use
```
$[coalesce(variables['system.pullRequest.sourceCommitId'], variables['build.sourceVersion'])]
```

**Note:** this approach doesn't populate the commit details for manually triggered runs. It is recommended that you provide the values for both branch and commit in this case.

> ---

Reviewer 2: ```suggestion
```

## Github Actions

Github Actions provides different event data based on if the build is triggered by a branch push versus via a Pull Request (PR). You need to use a conditional statement to select the right variable. Here's a suggested method to do so.

For *Git Reference*, use
```
${{ github.head_ref || github.ref }}
```

For *Git Commit*, use
```
${{ github.event.push.after || github.event.pull_request.head.sha }}
```

**Note:** this approach doesn't populate the commit details for manually triggered runs. It is recommended that you provide the values for both branch and commit in this case.

> ---

Reviewer 2: ```suggestion
```

## TeamCity

TeamCity provides different data based on if the build is triggered by a branch push versus via a Pull Request (PR). There is no easy way to automate selecting the right one. Based on your process, we suggest using one of the following options. 

### Git Reference

If the build is triggered by a branch push, use 
```
%teamcity.build.branch%
```
If the build is triggered via a PR, use 
```
%teamcity.pullRequest.source.branch%
```

### Git Commit

This is the same for both scenarios.
```
%build.vcs.number%
```

### Notes

> If you have multiple VCS sources used in a build, remember to add the fully qualified parameter that includes the VCS Root ID. For example, ```teamcity.build.branch.<VCS_Root_ID>``` and  ```build.vcs.number.<VCS_Root_ID>```

Reviewer 2: ```suggestion
If multiple VCS sources are used in a build, remember to add the fully qualified parameter that includes the VCS Root ID. For example, `teamcity.build.branch.<VCS_Root_ID>` and  `build.vcs.number.<VCS_Root_ID>`.
```

> For a build that is the result of a chained step, you may need to reference a parameter from the original build dependency. It will look something like this for a *Git Commit* - ```dep.<Dependant_Build_ID>.build.vcs.number``` or ```dep.<Dependant_Build_ID>.build.vcs.number.<VCS_Root_Id>```.

Reviewer 2: ```suggestion
For a build resulting from a chained step, you may need to reference a parameter from the original build dependency. It will look something like this for a *Git Commit* - `dep.<Dependant_Build_ID>.build.vcs.number` or `dep.<Dependant_Build_ID>.build.vcs.number.<VCS_Root_Id>`.
```
